nixos/modules/virtualisation/nixos-containers.nix

```
@@ -488,7 +488,9 @@ in
                       extraConfig = { options, ... }: {
                         _file = "module at ${__curPos.file}:${toString __curPos.line}";
                         config = {
-                          nixpkgs = if options.nixpkgs?hostPlatform && host.options.nixpkgs.hostPlatform.isDefined
+                          nixpkgs = if options.nixpkgs?pkgs && host.options.nixpkgs.pkgs.isDefined
+                                    then { inherit (host.config.nixpkgs) pkgs; }
+                                    else if options.nixpkgs?hostPlatform && host.options.nixpkgs.hostPlatform.isDefined
                                     then { inherit (host.config.nixpkgs) hostPlatform; }
                                     else { inherit (host.config.nixpkgs) localSystem; }
```

These lines were supposed to be a complete config (and a modern one if the host has `hostPlatform`).
The error you're getting originates from the `nixpkgs.system` default, in the container by the looks of it.
What if we did:

```suggestion
                                    then { inherit (host.config.nixpkgs) hostPlatform; }
                                    else { inherit (host.config.nixpkgs) localSystem;
                                              inherit (host.config.nixpkgs.localSystem) system; }
```

Would that solve the problem?

It doesn't solve the problem that `nixpkgs.system` is read by something, which could be considered a root cause. That something should probably use `pkgs.hostPlatform` or `pkgs.hostPlatform.system` instead, and the error message should describe that possibility.

---

That doesn't solve the problem because none of `host.config.nixpkgs.{hostPlatform,localSystem,system}` are set.  Only `host.config.nixpkgs.pkgs` is set.

However, I *think* what might work is something like

```nix
if options.nixpkgs?pkgs && host.options.nixpkgs.pkgs.isDefined
then { localSystem = host.config.nixpkgs.pkgs.stdenv.system; }
else if …
```

… or something similar

---

Oh that makes sense. I was thinking about the container, but the problem is right here.

 `pkgs` is the best source of truth actually, so why not use it here too.

```suggestion
                                    then { inherit (host.pkgs.stdenv) hostPlatform; }
                                    else { localSystem = host.pkgs.stdenv.hostPlatform; }
```

---

How about just:

```nix
nixpkgs = { inherit (host.pkgs.stdenv) hostPlatform; };
```

---

That's the goal. I think we had that originally, but it was causing an issue for someone.
That was very early on and the new platform flavored options have proven themselves a bit, so maybe it's time to move all containers over to the new syntax.
That does increase the scope a bit, so I'd recommend to do that in a follow-up.

---

Oh do you mean that the container might not have a `nixpkgs.hostPlatform` option even if the host does?